chronos/achannels.nim

```diff
@@ -0,0 +1,254 @@
+  result.mask = -1
+  if maxItems <= 0:
+    result.maxItems = -1
+  else:
+    result.maxItems = maxItems
+  result.count = 0
+  result.wr = 0
+  result.rd = 0
+  result.refCount = 0
+  result.data = cast[ptr UncheckedArray[byte]](0)
+  result.initLocks()
+
+proc raiseChannelClosed() {.inline.} =
+  var err = newException(AsyncChannelError, "Channel closed or not opened")
+  raise err
+
+proc raiseChannelFailed() {.inline.} =
+  var err = newException(AsyncChannelError, "Channel synchronization failed")
+  raise err
+
+proc open*[Msg](chan: AsyncChannel[Msg]) =
+  ## Open channel ``chan``.
+  let loop = getGlobalDispatcher()
+  chan.acquireLock()
+  inc(chan.refCount)
+  chan.releaseLock()
+
+proc close*[Msg](chan: AsyncChannel[Msg]) =
+  ## Close channel ``chan``.
+  chan.acquireLock()
+  if chan.refCount == 0:
+    if not(isNil(chan.data)):
+      deallocShared(cast[pointer](chan.data))
+    chan.deinitLocks()
+    deallocShared(cast[pointer](chan))
+  else:
+    dec(chan.refCount)
+  chan.releaseLock()
+
+proc `$`*[Msg](chan: AsyncChannel[Msg]): string =
+  ## Dump channel ``chan`` debugging information as string.
+  chan.acquireLock()
+  result = "channel 0x" & toHex(cast[uint](chan)) & " ("
+  result.add("eventNotEmpty = 0x" & toHex(cast[uint](chan.eventNotEmpty)))
+  result.add(", eventNotFull = 0x" & toHex(cast[uint](chan.eventNotFull)))
+  result.add(", rd = " & $chan.rd)
+  result.add(", wr = " & $chan.wr)
+  result.add(", count = " & $chan.count)
+  result.add(", mask = 0x" & toHex(chan.mask))
+  result.add(", data = 0x" & toHex(cast[uint](chan.data)))
+  result.add(", maxItems = " & $chan.maxItems)
+  result.add(", refCount = " & $chan.refCount)
+  result.add(")")
+  chan.releaseLock()
+
+proc rawSend(rchan: RawAsyncChannel, pbytes: pointer, nbytes: int) =
+  var cap = rchan.mask + 1
+  if rchan.count >= cap:
+    if cap == 0: cap = 1
+    var n = cast[ptr UncheckedArray[byte]](allocShared0(cap * 2 * nbytes))
+    var z = 0
+    var i = rchan.rd
+    var c = rchan.count
+
+    while c > 0:
+      dec(c)
+      copyMem(addr(n[z * nbytes]), addr(rchan.data[i * nbytes]), nbytes)
+      i = (i + 1) and rchan.mask
+      inc(z)
+
+    if not isNil(rchan.data):
+      deallocShared(rchan.data)
+
+    rchan.data = n
+    rchan.mask = (cap * 2) - 1
+    rchan.wr = rchan.count
+    rchan.rd = 0
+
+  copyMem(addr(rchan.data[rchan.wr * nbytes]), pbytes, nbytes)
+  inc(rchan.count)
+  rchan.wr = (rchan.wr + 1) and rchan.mask
+
+proc send*[Msg](chan: AsyncChannel[Msg], msg: Msg) {.async.} =
+  ## Send message ``msg`` over channel ``chan``. This procedure will wait if
+  ## internal channel queue is full.
+  chan.acquireLock()
+  try:
+    if chan.refCount == 0:
+      raiseChannelClosed()
+
+    if chan.maxItems > 0:
+      # Wait until count is less then `maxItems`.
+      while chan.count >= chan.maxItems:
+        chan.releaseLock()
+        let res = await chan.eventNotFull.wait(InfiniteDuration)
```

Comment:
Are we setting ourselves up for a deadlock here (or maybe just an avoidable delay)? We want to wait until there's room in the channel's internal queue, so we can add our Msg to it, right? Looks like we're actually waiting for another channel operation to take a look at this queue's size and trigger this event for us (I see `recv()` doing that).

What if there are no other operations on this channel while we wait to send something through it? We want to be able to send something over a buffered channel before `recv()` is called on the other end, don't we?

Reply:
When Queue has limited size, it is possible that Queue will be full, in such way `send` will wait for reader to consume some data from Queue, so it can store new data to the Queue. If Queue has unlimited size this part of code will not be executed.